### PR TITLE
makefile: automatically update/use latest v.c and cache

### PR DESCRIPTION
**Pull request by zaoqi**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,7 +1,7 @@
 CC ?= cc
 CFLAGS ?= -O2 -fPIC
 PREFIX ?= /usr/local
-VC ?= 0.1.17
+VC ?= e0c9a84
```
Comment (unknown-v) [inline Makefile]: why not using `git rev-parse --short HEAD` to get the short hash of the current local head

Comment (zaoqi) [reply Makefile]: vc is another repository

Comment (unknown-v) [reply Makefile]: true,but we can use `something like 
`git ls-remote https://github.com/vlang/vc.git --tags master | awk '{print $1;}'`

Comment (zaoqi) [reply Makefile]: Specifying a version can improve reproducibility.

Comment (unknown-v) [reply Makefile]: but the hash will change after this commit got marged

Comment (zaoqi) [reply Makefile]: but vc is another repository, so hash will not change

Comment (unknown-v) [reply Makefile]: ohh ,ok sry was a long day

Comment (medvednikov) [reply Makefile]: manually updating makefile every time vc is changed is not an option

File: Makefile [hunk 2]
```diff
@@ -1,7 +1,7 @@
 CC ?= cc
 CFLAGS ?= -O2 -fPIC
 PREFIX ?= /usr/local
-VC ?= 0.1.17
+VC ?= $(shell git ls-remote https://github.com/vlang/vc.git --tags master | awk '{print $$1;}')
```
Comment (zaoqi) [inline Makefile]: @medvednikov
now Makefile will automatically update/use optimized latest version of vc and cache

Comment (whoizit) [reply Makefile]: Is this valid?
```
$ git ls-remote https://github.com/vlang/vc.git --tags master | awk '{print $$1;}'
e0c9a842b3d034aa5e5a1e2ea8b22dc9464b7765  refs/heads/master
```

Comment (spytheman) [reply Makefile]: It is. In Makefiles $$ => $, so the awk will see $1 .
retrieve remote master commit hash inside Makefile

Comment (zaoqi) [reply Makefile]: @whoizit In Makefile, `$$` is `$`

Comment (spytheman) [reply Makefile]: A possible negative cost is that will force make to do a network operation on each make run, which may take +1,2 seconds.

**Comment by Rendims** [comment]
As it is now it doesn't actually save the latest v.c hash anywhere?
Meaning that future old commits still won't be reproducible.

How about saving `VC` to a file when v.c is downloaded which can later be committed together with any other changes?
```make
v.%.c:
	curl -o v.$*.c -LsSf https://github.com/vlang/vc/raw/$*/v.c
	echo -n $* > version_vc.txt
```

Then development is done with v.c from master but the reference is saved and old commits can be rebuilt using `VC=$(<version_vc.txt) make`.

File: Makefile (hunk 2 above)
Comment (Rendims) [reply Makefile]: That network operation can be avoided by manually setting a value to VC.
`VC=0.1.17 make` or `VC=master make`.
But that will of course skip any updates as well.

**Comment by zaoqi** [comment]
@Rendims 
or `update-vc`: https://github.com/vlang/v/blob/e7d8b69767c3832ef5612bfaa7b258ab177e9fdf/Makefile#L9

**Comment by spytheman** [comment]
I think storing just the hash in a separate .txt file is better than a self modifying Makefile.

**Comment by Rendims** [comment]
@zaoqi, `update-vc` could work and I like that it's a bit more explicit (but self modifying stuff is scary IMO).
Then skipping network operations (updates) and rebuilding old commits could be done with `make v`.

A separate file can be reused by the Windows batch files. But my initial suggestion, which implicitly updates the file, does however have the confusing behavior that `make && VC=oldhash make && make` stores `oldhash` in the file since the latest v.c is already downloaded.

**Comment by zaoqi** [comment]
@medvednikov @Rendims, merging vc and v can avoid these issues.

**Comment by spytheman** [comment]
Could not the vc repository be just a git submodule inside v repository?

In this way, updating vc (which could be done by `v up`), will rely on git, no need for curl at all.

Also having vc as a local submodule, will make the hash of the v.c commit available without network operations, or external storing of .txt files.

```shell
0[10:14:34] /v/v $ 
0[10:14:36] /v/v $ git clone https://github.com/vlang/v /v/v_with_vc_as_submodule/
Cloning into '/v/v_with_vc_as_submodule'...
remote: Enumerating objects: 22, done.
remote: Counting objects: 100% (22/22), done.
remote: Compressing objects: 100% (19/19), done.
remote: Total 5944 (delta 8), reused 8 (delta 3), pack-reused 5922
Receiving objects: 100% (5944/5944), 2.94 MiB | 1.14 MiB/s, done.
Resolving deltas: 100% (3833/3833), done.
0[10:15:05] /v/v $ cd /v/v_with_vc_as_submodule
0[10:15:07] /v/v_with_vc_as_submodule $ git submodule add https://github.com/vlang/vc vc
Cloning into '/v/v_with_vc_as_submodule/vc'...
remote: Enumerating objects: 49, done.
remote: Counting objects: 100% (49/49), done.
remote: Compressing objects: 100% (36/36), done.
remote: Total 308 (delta 15), reused 46 (delta 13), pack-reused 259
Receiving objects: 100% (308/308), 806.72 KiB | 1.19 MiB/s, done.
Resolving deltas: 100% (96/96), done.
0[10:15:30] /v/v_with_vc_as_submodule $ ls -lart vc/
total 472
-rw-rw-r--  1 delian delian     27 авг 15 10:15 .git
-rw-rw-r--  1 delian delian    115 авг 15 10:15 README.md
drwxrwxr-x  2 delian delian   4096 авг 15 10:15 .
-rw-rw-r--  1 delian delian 463701 авг 15 10:15 v.c
drwxrwxr-x 10 delian delian   4096 авг 15 10:15 ..
0[10:15:34] /v/v_with_vc_as_submodule $ git status
On branch master
Your branch is up to date with 'origin/master'.

Changes to be committed:
  (use "git reset HEAD <file>..." to unstage)

        new file:   .gitmodules
        new file:   vc

0[10:15:39] /v/v_with_vc_as_submodule $ cat .gitmodules 
[submodule "vc"]
        path = vc
        url = https://github.com/vlang/vc
0[10:15:44] /v/v_with_vc_as_submodule $ git add .                                                                                                             
0[10:15:53] /v/v_with_vc_as_submodule $ git commit -m "Add vc/ repository as git submodule."
[master 1cbe9ba] Add vc/ repository as git submodule.                                                                                                         
 2 files changed, 4 insertions(+)
 create mode 100644 .gitmodules
 create mode 160000 vc
0[10:16:06] /v/v_with_vc_as_submodule $ git ls-remote vc/ --tags master|head -n1
98ddb5b64db55bcf5515665be1566f0894498d23        refs/heads/master
0[10:16:07] /v/v_with_vc_as_submodule $ 
```

**Comment by spytheman** [comment]
Forcing updating of the vc submodule (possibly done by `v up`):
```
0[10:29:36] /v/v_with_vc_as_submodule $ git submodule update -f
Submodule path 'vc': checked out '98ddb5b64db55bcf5515665be1566f0894498d23'
0[10:29:37] /v/v_with_vc_as_submodule $
```

**Comment by spytheman** [comment]
Getting vc/ commit hash, using only git (no other unix tools needed, so suitable to be put in .bat files too):
```
0[10:44:58] /v/v_with_vc_as_submodule $ cd vc ; git rev-parse    HEAD ; cd ..
98ddb5b64db55bcf5515665be1566f0894498d23
0[10:45:01] /v/v_with_vc_as_submodule $ cd vc ; git rev-parse --short    HEAD ; cd ..
98ddb5b
0[10:45:11] /v/v_with_vc_as_submodule $ 
```

**Comment by medvednikov** [comment]
I've simplified Makefile and removed curl dependency.

**Comment by zaoqi** [comment]
@medvednikov
A negative cost is that will force make `git clone` and compile v.c on each make run, which may take +n seconds.

this method may be worse than the original https://github.com/vlang/v/blob/67d6702c1f0beee0acf5fd197a7b307002ff6de6/Makefile

**Comment by medvednikov** [comment]
@zaoqi make is supposed to run once. `v up` afterwards.

**Pull request closed without merging**